xc7/fasm2bels/make_routes.py

```diff
@@ -154,7 +154,10 @@ def add_node(
         self.parent_nodes[node_pkey] = parent_node_pkey
 
         for wire_pkey in get_wires_in_node(conn, node_pkey):
-            net_map[wire_pkey] = self.source_wire_pkey
+            if wire_pkey not in net_map:
+                net_map[wire_pkey] = set()
```

Comment:
I see this pattern a lot. What do you think about using this class for dictionaries of sets?

```python
class SetDict(dict):
    def __missing__(self, key):
        return set()
```

Reply:
I'm not a fan of this kind of magic, because it masks invalid keys down the road, and there isn't a strong readability improvement at this site because you still need to know that net_map is a `SetDict`, rather than just `dict`.

Also this strategy doesn't work because the return is to an object not in the dictionary itself.

```
class SetDict(dict):
    def __missing__(self, key):
        o = set()
        self[key] = o
        return o
```

would work, but again, this is bad pattern IMO.

Reply:
FYI Python also has the `defaultdict` object...
```python
from collections import defaultdict
mydict = defaultdict(set)
```

Reply:
BTW I'm not a huge fan of defaultdict and agree with @litghost's comment that it masks invalid keys.